### ISSUE TITLE
[data grid] singleSelect cell should display `option.label` instead of `option.value` when possible

### ISSUE THREAD

**Issue by juliezhang0225** [mui/mui-x#4437]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

I'm using MUI Data Grid column type - 'singleSelect'. And I'm using an object valueOptions to define the selection options. The table is showing the actual value instead of the label. I'm following the [this instruction](https://mui.com/zh/x/react-data-grid/columns/#special-properties). I also tried to change the value type from number to string, and it's still not working. Here is my code snippet: 

```
import * as React from 'react';
import { DataGrid, GridColumns, GridRowsProp } from '@mui/x-data-grid';


export default function BasicEditingGrid() {
    return (
        <div style={{ height: 300, width: '100%' }}>
            <DataGrid
                rows={rows}
                columns={columns}
                experimentalFeatures={{ newEditingApi: true }}
            />
        </div>
    );
}

const columns: GridColumns = [
    { field: 'program', headerName: 'Program', width: 180, editable: false },
    { 
        field: 'enrollment', 
        headerName: 'Enrollment', 
        width: 180, 
        type: 'singleSelect',
        valueOptions: [
            { value: 99, label: 'Select One' },
            { value: 1, label: 'Enrollment' },
            { value: 0, label: 'No Enrollment' }
        ],
        editable: true 
    },
];

const rows: GridRowsProp = [
    {
        id: 1,
        program: "abc",
        enrollment: 1,
    },
    {
        id: 2,
        program: "def",
        enrollment: 99,
    },
];
```
<img width="375" alt="React_App" src="https://user-images.githubusercontent.com/60940569/162643525-73a4d9ae-879d-4a5d-ab9a-67c499af5ffa.png">

### Expected behavior 🤔

The table should show the label instead of the actual value.

### Steps to reproduce 🕹

Steps:

1.
2.
3.
4.

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by alexfauquette** [comment]
I agree, to be consistent with filtering and editing the label property should be used by default.

The time we fix this bug, you can use the following workaround

``jsx
valueFormatter: ({ id: rowId, value, field, api }) => {
  const colDef = api.getColumn(field);
  const option = colDef.valueOptions.find(
    ({ value: optionValue }) => value === optionValue
  );

  return option.label;
},
```
 https://codesandbox.io/s/columntypesgrid-material-demo-forked-hmb3e9?file=/demo.js:715-967

**Comment by juliezhang0225** [comment]
Thanks! That works.

**Comment by alexfauquette** [comment]
I reopen the issue such that we can track this bug

**Comment by Algie37** [comment]
Can't get this workaround to run in Typescript. Does it have a typescript version?

**Comment by alexfauquette** [comment]
@Algie37 can you share a codesandbox with the workaround in typescript? I don't see how typescript could impact this workaround

**Comment by Algie37** [comment]
I dont understand. It works perfectly in Codesandbox {https://codesandbox.io/s/react-typescript-forked-wd7ky6?file=/src/DataGrid.tsx}. Howerver, on VS Code, It gives me the following Error on the [optionValue] item. 
![2022-07-26 02_54_50-DataTable tsx - mortman - Visual Studio Code](https://user-images.githubusercontent.com/16317334/180894100-1c65e55e-ba07-406e-8a05-326ca4b86017.png)

**Comment by alexfauquette** [comment]
@Algie37 you forget to save your codesandbox. But not a problem, this error depends on your typescript configuration. Do you allow implicit any or not.

Since we do not know the type of the value (could be any thing) we set any, On your side, you can specify the type of the `value` property
```ts
export type ValueOptions = string | number | { value: any; label: string };
```

**Comment by chaiwa-berian** [comment]
@alexfauquette While on this same issue, I am getting the error below both on my local machine and in the sandbox, would you know why? [See sandbox here](https://codesandbox.io/s/floral-river-3u7459?file=/src/App.tsx):
```
MUI: Could not find the data grid context.
It looks like you rendered your component outside of a DataGrid, DataGridPro or DataGridPremium parent component.
This can also happen if you are bundling multiple versions of the data grid.
```

**Comment by chaiwa-berian** [comment]
When `valueFormatter` is defined like this, it seems to work but fails when `useGridApiContext` is used. 
```
  valueFormatter: ({ value, field, api }) => {
        // This does not work but there is a deprecation message on api
        // const colDef = apiContext.current.getColumn(field);

        const colDef = api.getColumn(field);

        const option = colDef.valueOptions
          ? colDef.valueOptions.find((params: any) => value === params.value)
          : {};

        return option?.label;
      },
```

**Comment by alexfauquette** [comment]
You may be looking for `useApiRef` instead of `useApiContext`. More detail in [the doc ](https://mui.com/x/react-data-grid/api-object/)

**Comment by chaiwa-berian** [comment]
> You may be looking for `useApiRef` instead of `useApiContext`. More detail in [the doc ](https://mui.com/x/react-data-grid/api-object/)

Is `useApiRef` only available for `Pro` version? Passing `apiRef` to `DataGrid` gives the below error. It seems this is not  available for community version `@mui/x-data-grid`:
```
Property 'apiRef' does not exist on type 'IntrinsicAttributes & Omit<Partial<DataGridPropsWithDefaultValues> & DataGridPropsWithComplexDefaultValueBeforeProcessing & DataGridPropsWithoutDefaultValue<...>, DataGridForcedPropsKey> & { ...; } & RefAttributes<...>'
```

**Comment by alexfauquette** [comment]
Yes, little blue square means it's a pro feature

![image](https://user-images.githubusercontent.com/45398769/181454295-f755ecc6-54fc-4cd0-84d9-4809a50dbd6f.png)

But this is out of the topic of this issue. To keep this issue clean please open a new one if you need more information

**Comment by Algie37** [comment]
@alexfauquette Thank you for this solution Alex. I allowed implicit any and the error cleared. Works like a charm!
![2022-07-30 08_23_37-tsconfig json - mortman - Visual Studio Code](https://user-images.githubusercontent.com/16317334/181876140-eb6116f2-d532-44e9-81de-468de69a24a4.png)

**Comment by tiagoitferreira** [comment]
![image](https://user-images.githubusercontent.com/81890001/182257214-adfeb09f-ddae-4f8c-8fc9-2112827b335a.png)

**Comment by yaredtsy** [comment]
I think its simpler to use `valueGetter`
```javascript
valueGetter: ({ value, colDef }) => {
    const option = colDef.valueOptions.find(
      ({ value: optionValue }) => value === optionValue
    );
    return option.label;
  },
```

**Comment by cheryl-carpenter** [comment]
> ```js
> valueGetter: ({ value, colDef }) => {
>     const option = colDef.valueOptions.find(
>       ({ value: optionValue }) => value === optionValue
>     );
>     return option.label;
>   },
> ```

Can anyone show this in context please? I'm new to typescript and while I'd found the type definition in the mui packages, I don't get how to apply it to the same snippet referenced by @Algie37

**Comment by cheryl-carpenter** [comment]
> 



> ```ts
> export type ValueOptions = string | number | { value: any; label: string };
> ```

Can anyone show this in context please? I'm new to typescript and while I'd found the type definition in the mui packages, I don't get how to apply it to the same snippet referenced by @Algie37

**Comment by yaredtsy** [comment]
hey @PttRulez 
I believe you should be using `valueGetter` instead of `valueFormatter`. For more information, please refer to the documentation available here  https://mui.com/x/react-data-grid/column-definition/#value-getter.
you can implement like this.
```jsx
      valueGetter: ({ value, colDef }) => {
          console.log(colDef);
          const option = colDef.valueOptions.find(
            ({ value: optionValue }) => value === optionValue
          );
          if (option) return option.label;
        },
```

**Comment by yaredtsy** [comment]
Oh, I apologize for overlooking the details in the documentation. The filtering also isn't working when I use `valueGetter`.

![image](https://user-images.githubusercontent.com/34584266/213882404-8aa00205-5648-462d-b59f-d4ff25a78518.png)

Shouldn't this section be highlighted like the rest?

![image](https://user-images.githubusercontent.com/34584266/213882743-0cce1383-f750-4060-952f-c86f134476a6.png)
